src/components/tl/mlx5/mcast/tl_mlx5_mcast_allgather.c

```diff
@@ -0,0 +1,439 @@
+    req->to_send    = req->num_packets;
+    req->to_recv    = comm->commsize * req->num_packets;
+
+    comm->allgather_comm.coll_counter++;
+    return UCC_OK;
+}
+
+static inline ucc_status_t ucc_tl_mlx5_mcast_coll_do_allgather(void* sbuf, void *rbuf, int size,
+                                                               ucc_tl_mlx5_mcast_coll_comm_t *comm,
+                                                               ucc_tl_mlx5_mcast_coll_req_t **task_req_handle)
+{
+    ucc_tl_mlx5_mcast_coll_req_t *req;
+    ucc_status_t                  status;
+
+    tl_trace(comm->lib, "MCAST allgather start, sbuf %p, rbuf %p, size %d, comm %d, "
+             "comm_size %d, counter %d",
+             sbuf, rbuf, size, comm->comm_id, comm->commsize, comm->allgather_comm.coll_counter);
+
+    req = ucc_calloc(1, sizeof(ucc_tl_mlx5_mcast_coll_req_t), "mcast_req");
+    if (!req) {
+        tl_error(comm->lib, "malloc failed");
+        return UCC_ERR_NO_MEMORY;
+    }
+
+    status = ucc_tl_mlx5_mcast_prepare_allgather(sbuf, rbuf, size, comm, req);
+    if (UCC_OK != status) {
+        tl_warn(comm->lib, "prepare mcast allgather failed");
+        ucc_free(req);
+        return status;
+    }
+
+    status = UCC_INPROGRESS;
+
+    *task_req_handle = req;
+
+    return status;
+}
+
+ucc_status_t ucc_tl_mlx5_mcast_allgather_start(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_mlx5_task_t            *task      = ucc_derived_of(coll_task, ucc_tl_mlx5_task_t);
+    ucc_tl_mlx5_team_t            *mlx5_team = TASK_TEAM(task);
+    ucc_tl_mlx5_mcast_team_t      *team      = mlx5_team->mcast;
+    ucc_coll_args_t               *args      = &TASK_ARGS(task);
+    ucc_datatype_t                 dt        = args->src.info.datatype;
+    size_t                         count     = args->src.info.count;
+    ucc_status_t                   status    = UCC_OK;
+    size_t                         data_size = ucc_dt_size(dt) * count;
+    void                          *sbuf      = args->src.info.buffer;
+    void                          *rbuf      = args->dst.info.buffer;
+    ucc_tl_mlx5_mcast_coll_comm_t *comm      = team->mcast_comm;
+
+    task->coll_mcast.req_handle = NULL;
+
+    status = ucc_tl_mlx5_mcast_coll_do_allgather(sbuf, rbuf, data_size, comm, &task->coll_mcast.req_handle);
+    if (status < 0) {
+        tl_warn(UCC_TASK_LIB(task), "do mcast allgather failed:%d", status);
+        coll_task->status = status;
+        return ucc_task_complete(coll_task);
+    }
+
+    coll_task->status = status;
+
+    return ucc_progress_queue_enqueue(UCC_TL_CORE_CTX(mlx5_team)->pq, &task->super);
+}
+
+void ucc_tl_mlx5_mcast_allgather_progress(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_mlx5_task_t           *task      = ucc_derived_of(coll_task, ucc_tl_mlx5_task_t);
+    ucc_tl_mlx5_mcast_coll_req_t *req       = task->coll_mcast.req_handle;
+    ucc_status_t                  status;
+
+    if (task->coll_mcast.req_handle != NULL) {
```

Comment:
this will act more as a sanity check. if the start function is not called yet, and progress gets called before start, then this check ensures we do not segfault.

Reply:
so IMO an assert is preferable here. Up to you